oeps/oep-0045/decisions/0001-tutor-as-replacement-for-edx-configuration.rst

```diff
@@ -0,0 +1,499 @@
+
+
+Conclusion of the evaluation
+............................
+
+
+Positives
+`````````
+
+At the time of writing, Tutor is the only existing alternative to ``edx/configuration`` with a maintainer that is active
+in the community. It ticks a lot of the boxes that qualify a good replacement:
+
+-  It can do single-server production deployments well
+-  It allows for automated deployment as well as, or better than, ``edx/configuration``
+-  It supports Koa, and new releases are added quickly
+-  It allows for theme customization, running edx-platform forks, and installing custom requirements
+-  It is extensible via plugins
+-  It can be used as a lightweight, efficient Open edX development environment
+-  Actively maintained by a committed developer; codebase has no technical debt
+-  Licensed under the AGPL v3
+
+
+Omissions
+`````````
+
+There is work to be done before it can be considered a full replacement:
+
+-  There must be a documented way to run highly-available, multi-server production deployments, including a way to run
+   rolling upgrades with as little downtime as possible (see note below)
+-  Migration from the native installation must be documented
+-  It must support deployments of the master branches of edx-platform and IDAs, and once that comes about it must
+   undergo CI/CD so that breakages can be dealt with quickly
+-  There should be more than one maintainer
+
+.. note::
+   The nature of Open Releases (Juniper, Koa, etc.) make it very difficult to allow for no-downtime rolling upgrades:
+   they're a mashup of multiple edx.org releases which over the relevant time period introduce backward-incompatible
+   migrations that can't be easily conflated.  It should, however, be possible to do if one is tracking master closely
+   enough.
+
+It is estimated that initial versions of the above will take anywhere from three to six months to achieve, if sufficient
+developers commit resources to the endeavor. (Part of the test of whether a community-supported deployment method will
+work hinges precisely on whether this community can pool resources to maintain it!)
+
+
+Wishlist
+````````
+
+The following would be great to have once the TODO list above is taken care of:
+
+-  Support the deployment of most MFEs
+-  There should be a way to reuse the same base LMS image with different themes without having to bake them into the
+   image
+-  A easier to set up local development environment
+
+
+Decisions
+---------
+
+1. Adopt Tutor as the community-maintained replacement for ``edx/configuration`` for production deployments, with the
+   knowledge that it needs to be `further developed <#omissions>`__ in order to achieve all the described `criteria
+   <#criteria>`__.
+
+2. For the following release, Lilac, ``edx/configuration`` will be supported but marked as deprecated, and Tutor will be
+   supported for the first time.
```

Comment:
@nedbat, you suggested adding this to the ADR.  I agree. :)  Does the wording work for you?  (See also the following decision.)

Reply:
How will users choose between Tutor and the native installation? Can we suggest that Tutor be the recommended installation method for new users, and that users who already run the native installation keep using it?

Reply:
One challenge with that approach is the question of what will the native installation be once the Ansible repository is fully deprecated?

Reply:
If the decision is provisional, and we would approach this as an "alpha" release, it would make sense to ask for feedback on it from users willing to become involved (i.e. post their comments about the Tutor approach) or who want to start looking into migrating their setups. If we directly recommend all new users to use Tutor, it makes it harder to take the final decision objectively, as we would already have redirected a significant part of the community to it.

Reply:
I don't have the time at the moment to commit to this, but it is on my roadmap to do a test deployment of Tutor in our QA environment over the next couple of months.

Reply:
> it would make sense to ask for feedback on it from users willing to become involved (i.e. post their comments about the Tutor approach)

I suggest making that survey on the [Tutor forums](https://discuss.overhang.io)... There are plenty of users there who are happily running Tutor in production, either with docker-compose or Kubernetes.

Reply:
@regisb Yes, I can imagine that the current Tutor users are happy with it - at least I would hope so :) My point here is to ask the larger Open edX community, who isn't using Tutor. The use cases might be different, we would need to test the upgrade path, etc. I'd assume everyone would be happy with the change - but that's something we need to confirm, and that "alpha" phase would be good for that?

Reply:
I think my issue is with the term: "alpha": https://en.wikipedia.org/wiki/Software_release_life_cycle#Alpha

> Alpha software is not thoroughly tested by the developer before it is released to customers. Alpha software may contain serious errors, and any resulting instability could cause crashes or data loss. Alpha software may not contain all of the features that are planned for the final version.

Tutor has been thoroughly tested and all "serious bugs" which have been reported so far have been fixed. To the best of my knowledge, Tutor does not cause crashes or data loss. Not all features from the native installation are available in Tutor (I'm thinking of Insights, for instance), but these features are explicitely not on the roadmap.: https://docs.tutor.overhang.io/faq.html#what-features-are-missing-from-tutor

Thus, the term "alpha software" does not apply to Tutor. So I'd like to propose the following recommendations:

- Existing users of the native installation should keep running the Ansible playbooks officially supported by edX.
- New users should use Tutor, which should be considered a "release candidate" by the Open edX community.

@blarghmatey: it is a prerequisite of this ADR that someone, at some point, will write migration instructions from the native installation to Tutor. BTW, I think it should be time to look for volunteers. Is this a task for the BTR working group?

Reply:
> One challenge with that approach is the question of what will the native installation be once the Ansible repository is fully deprecated?

I don't know that we want to support more than one installation approach.  If we like Tutor, it will eventually be the only approach, right?

BTW: "Native" originally meant, "directly on an Ubuntu box" as opposed to "with Vagrant".  We don't need to keep the word Native.

Reply:
@regisb The term alpha doesn't refer to the version of Tutor - but more of using it as the official installation of Open edX for everyone in the community. It has been extensively tested by people using it currently, but not by the rest of the community who would then have to adopt it -- and migrate to it.

Reply:
What you are describing here is not "alpha" software, but a release candidate: https://en.m.wikipedia.org/wiki/Software_release_life_cycle#Release_candidate

> A release candidate (RC), also known as "going silver", is a beta version with potential to be a stable product, which is ready to release unless significant bugs emerge. In this stage of product stabilization, all product features have been designed, coded and tested through one or more beta cycles with no known showstopper-class bugs.

Reply:
@regisb Release Candidate could be fine yep :) The main point is to make sure people realize they should only follow this path if they are ready to get involved in ironing out issues with us.